### PR TITLE
Add Humidity Control Feature with Automatic DRY Mode Switching

### PR DESCRIPTION
**Pull request by paulczar**
# Add Humidity Control Feature with Automatic DRY Mode Switching

## Problem Statement

In high-humidity climates and regions, air conditioning systems provide cooling that naturally dehumidifies the air while running. However, when the target temperature is reached and the AC stops running, humidity levels can quickly rise again, making the space feel uncomfortable despite the correct temperature being maintained.

**The Core Issue:**
- AC cooling naturally dehumidifies while actively running
- When temperature reaches target, the AC stops running
- Humidity can remain high (60%+ relative humidity)
- Space feels "muggy" or "sticky" even at correct temperature
- Users may unnecessarily lower temperature setpoint to keep AC running, wasting energy

**Why This Matters:**
- Human comfort depends on both temperature AND humidity levels
- High humidity reduces the effectiveness of evaporative cooling from skin
- Can lead to overcooling and significantly higher energy consumption
- Particularly problematic in tropical, subtropical, and coastal climates
- Common issue in basements and naturally humid environments

## Solution

This PR adds automatic humidity control that switches the AC to DRY mode when:
- Temperature is at or near target (cooling not actively needed)
- Humidity exceeds a configurable threshold (default: 60%)
- AC is in COOL mode
- The underlying AC device supports DRY mode

**Key Features:**
- **Automatic Mode Switching**: Seamlessly switches between COOL and DRY modes based on conditions
- **Cooling Priority**: COOL mode always overrides DRY when temperature control is needed
- **Energy Efficient**: DRY mode uses less energy than active cooling
- **Configurable Threshold**: Adjustable per thermostat (default 60%)
- **Sensor Flexibility**: Works with any humidity sensor or input_number entity

## How It Works

1. **Monitoring**: Continuously monitors humidity via configured sensor entity
2. **Threshold Check**: Compares current humidity to user-defined threshold
3. **Cooling Need Assessment**: Uses proportional algorithm's `on_percent <= 0.05` to determine if active cooling is needed
4. **Mode Switching**:
   - If humidity > threshold AND cooling not needed → switch to DRY mode
   - If temperature rises and cooling needed → switch back to COOL mode
5. **Priority System**: Temperature control always takes precedence over humidity control

## Technical Implementation

**New Components:**
- `FeatureHumidityManager`: Manages humidity monitoring and threshold checking
- Config flow integration: Humidity sensor selection and threshold configuration
- State manager integration: DRY mode switching logic in HVAC mode calculation
- Central configuration support: Shared threshold across multiple thermostats

**Integration Points:**
- Integrated into `BaseThermostat` following existing feature manager pattern
- State manager checks humidity after window/auto-start-stop but before final mode assignment
- Cooling override uses proportional algorithm's `on_percent` to determine active cooling need

**Configuration:**
- Enable in Features step: "Use humidity management"
- Configure in Humidity management step:
  - Select humidity sensor entity
  - Set humidity threshold (default: 60%)
  - Option to use central configuration

## Use Cases

1. **Tropical Climates**: Maintain comfort when temperature is at target but humidity is high
2. **Coastal Areas**: Address high ambient humidity levels
3. **Basements**: Control humidity in naturally damp spaces
4. **Energy Efficiency**: Use DRY mode instead of overcooling to reduce humidity

## Testing

- ✅ Unit tests for `FeatureHumidityManager` (10 test cases)
- ✅ State manager integration tests (8 scenarios)
- ✅ Covers: threshold checking, sensor changes, cooling override, edge cases
- ✅ All tests passing

## Documentation

- ✅ Feature documentation: `documentation/en/feature-humidity.md`
- ✅ Testing guide: `TESTING_HUMIDITY.md`
- ✅ Updated README and base-attributes documentation
- ✅ Translation support: English translations included

## Compatibility

- Only available for AC mode thermostats (`over_climate` type)
- Requires underlying AC device to support DRY mode
- Backward compatible: Existing configurations unaffected
- Optional feature: Must be explicitly enabled

## Benefits

1. **Improved Comfort**: Better comfort in high-humidity conditions without overcooling
2. **Energy Savings**: More efficient than running AC just for dehumidification
3. **Automatic Operation**: No manual intervention required
4. **Flexible Configuration**: Per-thermostat threshold configuration
5. **Maintains Priority**: Temperature control always takes precedence

---

This feature addresses a common comfort issue in humid climates where temperature alone doesn't ensure comfort. By automatically switching to DRY mode when appropriate, users can maintain comfort without overcooling, resulting in better comfort and energy efficiency.

### CODE REVIEW

**Comment by jmcollin78** [comment]
Wahou, very impressive ! I will need some time to analyse your great work. I'm currently on the 8.3 release and I have more work on it. Be patient please.

**Comment by KipK** [comment]
Great one.
Could the humidity value  of the added sensor be exposed to the climate entity over "current_humidity" attribute then?

As it's part of standard attribute: https://developers.home-assistant.io/docs/core/entity/climate/

**Comment by jmcollin78** [comment]
Hello @paulczar, thank you for this great PR.
I'm currently reviewing it.

When review will be complete please update translations for other locales. You can leave the text in English and translator contributor will translate into their native language which is much better than an automatic translation. (README and translations/).

**Comment by jmcollin78** [comment]
Your formatter breaks all long line and dont use my custom config in pyproject.toml (`line-length = 180`).
Maybe you don't install black formatter. This makes many unnecessary changes and a make the review not easy.

This is not visible in the github review window but visible in the files you have commited.

All changes in base_thermostat.py (many are false changes) :
Capture d’écran 2025-12-13 à 10 41 54

File: custom_components/versatile_thermostat/config_schema.py [hunk 1]
```diff
@@ -90,6 +90,7 @@
         vol.Optional(CONF_USE_MOTION_FEATURE, default=False): cv.boolean,
         vol.Optional(CONF_USE_POWER_FEATURE, default=False): cv.boolean,
         vol.Optional(CONF_USE_PRESENCE_FEATURE, default=False): cv.boolean,
+        vol.Optional(CONF_USE_HUMIDITY_FEATURE, default=False): cv.boolean,
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/config_schema.py:93]: Climate with direct valve regulation cannot dry. It is for thermostatic valve which can't cool.
Remove this

File: custom_components/versatile_thermostat/config_schema.py [hunk 2]
```diff
@@ -71,6 +69,7 @@
...
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/config_schema.py:72]: Non over_climate VTherm cannot Cool and dehumidify. It is valid only for over_climate vtherm.
Remove this.

File: custom_components/versatile_thermostat/base_thermostat.py [hunk 3]
```diff
@@ -184,16 +186,15 @@ def __init__(
...
-        )
+        self._presence_manager: FeaturePresenceManager = FeaturePresenceManager(self, hass)
         self._power_manager: FeaturePowerManager = FeaturePowerManager(self, hass)
         self._motion_manager: FeatureMotionManager = FeatureMotionManager(self, hass)
         self._window_manager: FeatureWindowManager = FeatureWindowManager(self, hass)
         self._safety_manager: FeatureSafetyManager = FeatureSafetyManager(self, hass)
         # Auto start/stop is only for over_climate
         self._auto_start_stop_manager: FeatureAutoStartStopManager | None = None
         self._lock_manager: FeatureLockManager = FeatureLockManager(self, hass)
+        self._humidity_manager: FeatureHumidityManager = FeatureHumidityManager(self, hass)
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/base_thermostat.py:197]: Should be added only over climate VTherm. So remove from here and set in thermostat_climate.py only (like auto-start/stop manager typically). Set it to None here (like auto_start_stop_manager)

File: custom_components/versatile_thermostat/base_thermostat.py [hunk 4]
```diff
@@ -202,6 +203,7 @@ def __init__(
         self.register_manager(self._safety_manager)
         self.register_manager(self._safety_manager)
         self.register_manager(self._lock_manager)
+        self.register_manager(self._humidity_manager)
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/base_thermostat.py:206]: same

File: custom_components/versatile_thermostat/base_thermostat.py [hunk 5]
```diff
@@ -1397,6 +1383,12 @@ async def async_control_heating(self, timestamp=None, force=False) -> bool:
         # check auto_window conditions
         await self._window_manager.manage_window_auto(in_cycle=True)
 
+        # Refresh humidity state
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/base_thermostat.py:1386]: idem. Should be only for `over_climate`.

File: custom_components/versatile_thermostat/config_schema.py
Comment (jmcollin78) [inline custom_components/versatile_thermostat/config_schema.py:1]: Formatter false changes

File: custom_components/versatile_thermostat/feature_humidity_manager.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                    self._humidity_sensor_changed,
+                )
+            )
+
+    @overrides
+    async def refresh_state(self) -> bool:
+        """Tries to get the last state from sensor
+        Returns True if a change has been made"""
+        ret = False
+        if self._is_configured and self._humidity_sensor_entity_id:
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/feature_humidity_manager.py:79]: ... and self.entity_id not necessaray because is_configured cannot be true else.

File: custom_components/versatile_thermostat/feature_humidity_manager.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FeatureHumidityManager(BaseFeatureManager):
+    """The implementation of the Humidity feature"""
+
+    unrecorded_attributes = frozenset(
+        {
+            "humidity_sensor_entity_id",
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/feature_humidity_manager.py:36]: Will not work because it is nested into `humidity_manager` attribute.
You have to unrecorde `humidity_manager`.

Add the `current_humidity` (if manager is defined) into the `specific_states` attributes of the base-thermostat.py

File: custom_components/versatile_thermostat/state_manager.py [hunk 8]
```diff
@@ -149,6 +149,28 @@ async def calculate_current_hvac_mode(self, vtherm: "BaseThermostat") -> bool:
             elif vtherm.vtherm_hvac_mode != VThermHvacMode_HEAT and VThermHvacMode_HEAT in vtherm.vtherm_hvac_modes:
                 self._current_state.set_hvac_mode(VThermHvacMode_HEAT)
 
+        # Check humidity control - switch to DRY mode when humidity is too high and cooling is not needed
+        elif vtherm.humidity_manager.is_configured and vtherm.ac_mode and self._requested_state.hvac_mode == VThermHvacMode_COOL and VThermHvacMode_DRY in vtherm.vtherm_hvac_modes:
+            # Check if cooling is actively needed
+            # If on_percent is low (<= 0.05), cooling is not needed and we can use DRY mode
+            # If on_percent is high (> 0.05), cooling is needed and COOL mode takes priority
+            cooling_needed = False
+            if vtherm.proportional_algorithm and vtherm.proportional_algorithm.on_percent is not None:
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/state_manager.py:158]: In `over_climate` there is no TPO algorithm so you cannot use it here. I see that in your documentation but this assumption is wrong. instead you could use the `accumulated_error` of the pi algorithm (`PITemperatureRegulator`) which is used by `over_climate` self regulation. Not sure that if the self regulation is not configured you will have any pi algo running. We will have to chek that.

Not sure it is a good idea finally because the result will depend on the self regulation level.

Maybe it is better to compare directly the current room and the target temperature. If there are at let say 90% then dry is possible.

This business rule should be in the humidity-manager and not in the state-manager.py which should be "dumb" as possible (bo business rules of feature manager).

File: custom_components/versatile_thermostat/state_manager.py [hunk 9]
```diff
@@ -149,6 +149,28 @@ async def calculate_current_hvac_mode(self, vtherm: "BaseThermostat") -> bool:
...
+
+            if not cooling_needed and vtherm.humidity_manager.is_humidity_too_high:
+                # Temperature is at target, but humidity is too high - switch to DRY mode
+                _LOGGER.info(
+                    "%s - Humidity too high (%.1f%% > %.1f%%), switching to DRY mode",
+                    vtherm,
+                    vtherm.humidity_manager.current_humidity,
+                    vtherm.humidity_manager.humidity_threshold,
+                )
+                self._current_state.set_hvac_mode(VThermHvacMode_DRY)
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/state_manager.py:169]: Add a message to the user when `current_state` is not `requested_state`. See how it is done for `vtherm.set_hvac_off_reason` or `vtherm.set_temperature_reason`.

I propose to generalize the `vtherm.set_hvac_off_reason`, rename it to `vtherm.set_hvac_reason` and add a new reason like `hvac_dry_humidity_too_high` (+ translations).

**Review by jmcollin78 — COMMENTED** [review]
Globally it is a very high quality PR. Thank you so much for this.
The main comment are on the VTherm's type it will apply to.
It can be used for `ovrer_climate` (only `climate` entity can have a DRY mode).
So the feature should be implemented only for `over_climate` and menu visible only for this type of VTherm. You have an example with `auto-start/stop` feature which is only possible for `over_climate`.

The other mistake is the use of TPI algo (`on_percent`) and we have to find another way. I put some idea in the comments.

**Comment by paulczar** [comment]
Thanks mate, it'll take me a few days (maybe the weekend) to go over your reviews and comments.

**Comment by jmcollin78** [comment]
Hello, very sorry but I have to rewrite the main branch. You have to rebase with the new main branch.

**Comment by jmcollin78** [comment]
Hello @paulczar ,

Sorry to urge you but the more you wait the more you will have complicated merge and conflicts. Code base is changing fast. Do you think you will be able to finish this PR ?

**Comment by jmcollin78** [comment]
Hello, you are now very late. If this PR is finally stall, I will have to close it.

**Comment by jmcollin78** [comment]
No news for you since almost 3 month yet. The code has change drastically now and this PR cannot be merged.

Thank you anywayn but don't propose PR if you are not able to follow it after. It is only a waste of time.

**Comment by jack338c** [comment]
I really need this function,it's very useful to me.I living in tropical climates and coastal areas
looking forward to geting good news from you later
